### DISCUSSION TITLE
Аутентификация ⚡️ Node.js с примерами кода

### DISCUSSION THREAD

**Discussion by giscus** [xsltdev/nodejsdev.ru#55] [Announcements]
# Аутентификация ⚡️ Node.js с примерами кода

Справочник Node.js с примерами кода

https://nodejsdev.ru/guides/webdraftt/authentication/

**Comment by Sakreed** [comment]
Я слишком тупой, я ничего не понял

**Comment by Bahurtsev** [comment]
А почему нужен именно этот паспорт? Почему я не могу сделать как в PHP? Например:

if(логинИзФормы == логинИзБазыДанных){
 Создаём сессию и что-то делаем.
}

Ведь я же могу так сделать и без паспорта. И видел на Ютубе так даже делают.

**Reply by bndby** [reply]
аутентификация может быть не только через форму, но и, например, через соцсети. Чтобы не реализовывать для каждой системы свою авторизацию, используют библиотеки, например этот паспорт

**Comment by Bahurtsev** [comment]
Мда, объяснили вы эту тему так себе. Я бы сказал это только 2% от того что надо знать. Вместо res.status(200).send(); я отправляю файл как ответ, потому что в реальной жизни мы это используем. И вот если правильно то страницу secret, если не правильно то страницу login. Я запарился пока разобрался где нужно прописать чтобы всё работало. Потому что порядок очень важен, а если не писать не по порядку, то одно не работает, то другое. Кстати походу модуль обновился и надо переделывать. new localStrategy() принимает теперь 4 параметра. Да и неправильно перекидывать пользователя обратно на страницу /login если он ввёл что-то не правильно, тем более он уже на ней находится. Надо было прописать, если пользователь ввёл неправильный логин, вернуть ему сообщение на страницу: "Неправильный логин", если неверный пароль, то вернуть ему сообщение "Неверный пароль". Блин, по всему интернету, по крупицам собираю части несуществующего туториала чтобы решить задачу. Нет чтобы было всё сразу понятно описано.

**Reply by bndby** [reply]
Это не продакшн код, это обучающий пример. Как и всякий учебник, со временем он устаревает.

**Comment by Bahurtsev** [comment]
А и тут ещё ошибка:

app.get('/home', checkAuth(), (req, res) => {

Функция chekAuth будет запускаться много раз, надо так:

app.get('/home', checkAuth, (req, res) => {

**Reply by NikolaiBelski** [reply]
https://www.passportjs.org/concepts/authentication/password/      это смотрел?

**Reply by steamcards** [reply]
Там сама функция не рабочая, с ней код в бесконечное ожидание уходит. В таком виде работает:
```
    function checkAuth(req, res, next) {
        if (req.user) return next();
        else res.redirect('/login');
    }
```
Вызов:
```
app.get('/home', checkAuth, (req, res) => {}
```

**Comment by YuraKissil** [comment]
С учётом ранее сделанных комментариев у меня (node v22.14.0) с такими установками по ранее пройденным материалам:
```json
{
  "name": "testing",
  "typeAAA": "module",
  "version": "0.0.1",
  "description": "testing node & express",
  "main": "app.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [
    "node",
    "npm",
    "express",
    "testing"
  ],
  "author": "yura",
  "license": "ISC",
  "dependencies": {
    "connect-flash": "^0.1.1",
    "connect-redis": "^9.0.0",
    "cookie-parser": "^1.4.7",
    "express": "^5.1.0",
    "express-handlebars": "^8.0.2",
    "express-session": "^1.18.2",
    "passport": "^0.7.0",
    "passport-local": "^1.0.0",
    "pug": "^3.0.3",
    "redis": "^5.8.3",
    "request": "^2.88.2"
  }
}
```
сработал следующий код:
```javascript
const express = require('express'),
    app = express(),
    session = require('express-session'),
    passport = require('passport'),
    localStrategy = require('passport-local').Strategy,
    flash = require('connect-flash');

const host = '127.0.0.1';
const port = 7000;

function checkAuth() {
    console.log('checkAuth()');
    return app.use((req, res, next) => {
        console.log('checkAuth() call');
        if (req.user) next();
        else res.redirect('/login');
    });
}
function checkAuth(req, res, next) {
    console.log('checkAuth() call');
    if (req.user) next();
    else res.redirect('/login');
}

app.use(express.static(`${__dirname}/assets`));

passport.serializeUser((user, done) => done(null, user));
passport.deserializeUser((user, done) => done(null, user));

app.use(express.json());
app.use(express.urlencoded({ extended: true }));
app.use(session({ secret: 'my_secret_key' }));
app.use(flash());
app.use(passport.initialize());
app.use(passport.session());

passport.use(
    new localStrategy((username, password, done) => {
        console.log('localStrategy username=' + username + ' password=' + password);
        if (username !== 'test_user')
            return done(null, false, {
                message: 'Username not found',
            });
        else if (password !== 'test_password')
            return done(null, false, {
                message: 'Wrong password',
            });

        return done(null, { id: 1, name: 'Test', age: 21 });
    })
);

app.get('/login', (req, res) => {
    console.log('app.get(/login)');
    res.send('Login page. Please, authorize.<form action="/login" method="post"><input type="text" name="username" value="" /><input type="password" name="password" value="" /><input type="submit" name="submit" value="submit" /></form>');
});

app.post(
    '/login',
    passport.authenticate('local', {
        successRedirect: '/home',
        failureRedirect: '/login',
        failureFlash: true,
    })
);

app.use(checkAuth);
/*
app.use((req, res, next) => {
    console.log('check B');
    if (req.user) next();
    else res.redirect('/login');
});
*/
app.get('/home', checkAuth, (req, res) => {
    res.send("Home page. You're authorized.");
});

app.listen(port, host, function () {
    console.log(`Server listens http://${host}:${port}`);
});
```
Доработки не велики, хвосты оставил для наглядности.
Для тех, кто собирает по крупицам, думаю пригодиться.

**Reply by YuraKissil** [reply]
можно убрать вызов
```javascript
app.use(checkAuth);
```
если нет желания защитить следующие после него адреса и проверять их каждый отдельно с помощью
```javascript
app.get('/home', checkAuth, (req, res) => {
```
А можно отказаться от индивидуальной проверки последним методом в пользу общей защиты по первому.
Кому как больше подходит.

**Reply by YuraKissil** [reply]
И ещё в попыхах забыл закомментировать первую 
```javascript
/*
function checkAuth() {
    console.log('checkAuth()');
    return app.use((req, res, next) => {
        console.log('checkAuth() call');
        if (req.user) next();
        else res.redirect('/login');
    });
}
*/
```
